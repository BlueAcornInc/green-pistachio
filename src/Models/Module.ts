type ModuleConstructorArgs = {
    sourceDirectory: string;
    name: string;
};

export default class Module {
    private name: string;
    private sourceDirectory: string;

    constructor(config: ModuleConstructorArgs) {
        const { name, sourceDirectory } = config;

        this.name = name;
        this.sourceDirectory = sourceDirectory;
    }

    public getName() {
        return this.name;
    }

    public getSourceDirectory() {
        return this.sourceDirectory;
    }
}