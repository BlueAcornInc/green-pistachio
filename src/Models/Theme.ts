import glob from 'fast-glob';
import debug from 'debug';
const logger = debug('gpc:theme');

type ThemeConstructorArgs = ThemeData & {
    sourceDirectory: string;
    parent?: Theme;
};

// type CriticalCssInput = {
//     filepath: string;
//     urls: string[];
// };

export type ThemeData = {
    area: "frontend" | "adminhtml";
    path: string;
    enabled?: boolean;
    locales?: string[];
    stylesheets?: string[];
    // criticalCss?: CriticalCssInput[];
};

export default class Theme {
    private sourceDirectory: string;
    private parent?: Theme;
    private enabled: boolean | undefined;
    private data: ThemeData;

    constructor(config: ThemeConstructorArgs) {
        const {
            sourceDirectory,
            area,
            path,
            parent,
            locales,
            stylesheets,
            // criticalCss,
            enabled
        } = config;

        this.sourceDirectory = sourceDirectory;
        this.parent = parent;
        this.enabled = enabled || false;
        this.data = {
            area,
            path,
            locales,
            stylesheets,
            // criticalCss
        };
    }

    public configure(data: Partial<ThemeData>) {
        this.data = {
            ...this.data,
            ...data
        };
    }

    public getSourceDirectory() {
        return this.sourceDirectory;
    }

    public getParent() {
        return this.parent;
    }

    public getData() {
        return this.data;
    }

    public getLocales(): string[] {
        if (this.data.locales) {
            return this.data.locales;
        }

        logger(`${this.data.path}: Can't Determine Locales, defaulting to en_US`);

        return ['en_US'];
    }

    public getStyleSheets(): string[] {
        if (this.data.stylesheets) {
            return this.data.stylesheets;
        }

        logger(`${this.data.path}: Stylesheets not explicitly defined, detecting entries now.`);

        const files: Set<string> = new Set();

        const discoveredFiles = glob
            .sync(`${this.getSourceDirectory()}/**/[^_]*.less`)
            .map(sourceFile => sourceFile.replace(`${this.getSourceDirectory()}/web/`, ''));

        for (const file of discoveredFiles) {
            files.add(file);
        }

        let childTheme = this.getParent();

        while (childTheme) {
            for (const file of childTheme.getStyleSheets()) {
                files.add(file);
            }
            childTheme = childTheme.getParent();
        }

        const stylesheets = Array.from(files);

        logger(`${this.data.path}: Discovered the following stylesheets: ${stylesheets.join(', ')}`);

        return stylesheets;
    }

    // public getCriticalPaths(): CriticalCssInput[] {
    //     if (this.data.criticalCss) {
    //         return this.data.criticalCss;
    //     }

    //     logger(`No critical path data available for: ${this.data.path}`);

    //     return [];
    // }

    public getEnabled() {
        return this.enabled;
    }

    public setEnabled(enabled: undefined | boolean) {
        this.enabled = enabled;
    }
}
