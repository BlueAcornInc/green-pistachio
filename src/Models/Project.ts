import { join } from 'path';
import { SyncHook } from 'tapable';
import debug from 'debug';
import Module from "./Module";
import Theme, { ThemeData } from "./Theme";
import ConfigLoader from './Project/ConfigLoader';
import { Configuration } from 'webpack';
import WebpackConfigFactory from '../Webpack/ConfigFactory';
const logger = debug('gpc:project');

type ProjectConstructorArgs = {
    root: string;
    themes: Theme[];
    modules: Module[];
    enabledThemes?: string[];
};

type WebpackDevServerConfig = {
    proxyUrl: string;
    devServerUrl: string;
    devServerPort: number;
};

export default class Project {
    public root: string;
    private themes: Theme[];
    private enabledThemes: Theme[];
    private modules: Module[];
    private webpackDevelopmentMode: boolean = false;
    private proxyUrl: string = '';
    private devServerConfig: Partial<WebpackDevServerConfig> = {};

    public hooks = {
        // These hooks allow for modifying various gulp configurations
        gulp: {
            babelConfig: new SyncHook(["config"]),
            babelTypeScriptConfig: new SyncHook(["config"]),
            pngSpriteConfig: new SyncHook(["config"]),
            svgSpriteConfig: new SyncHook(["config"]),
            eslintConfig: new SyncHook(["config"]),
            criticalCssConfig: new SyncHook(["config"]),
        },
        webpack: {
            config: new SyncHook(["config"]),
            entryGlobs: new SyncHook(["globs"]),
            appendConfig: new SyncHook<[Configuration[], WebpackConfigFactory]>(["webpackConfigs", "configFactory"])
        },
        configure: new SyncHook(["project"])
    };

    public experiments = {
        webpack: {
            // Set to true in the the project.configure hook to enable unstable HMR functionality
            hmr: false,
            // Set to true to enable css modules
            cssModules: false,
            // Set to true to emit files to the pub directory instead of the individual theme directories
            emitFilesToPub: false,
        }
    };

    constructor(config: ProjectConstructorArgs) {
        const { root, themes, modules, enabledThemes } = config;

        this.root = root;
        this.themes = themes;
        this.modules = modules;
        this.enabledThemes = enabledThemes
            ? themes.filter(theme => enabledThemes.includes(theme.getData().path))
            : [];
    }

    public configure() {
        if (this.enabledThemes.length > 0) {
            // Enable default themes
            for (const theme of this.enabledThemes) {
                theme.setEnabled(true);
            }
        } else {
            for (const theme of this.themes) {
                if (
                    !theme
                        .getSourceDirectory()
                        .includes(
                            join(
                                this.getRootDirectory(),
                                'vendor'
                            )
                        )
                ) {
                    theme.setEnabled(true);
                    this.enabledThemes.push(theme);
                }
            }
        }

        // Enable default modules
        for (const magentoModule of this.modules) {
            if (
                !magentoModule
                    .getSourceDirectory()
                    .includes(
                        join(
                            this.getRootDirectory(),
                            'vendor'
                        )
                    )
            ) {
                magentoModule.setEnabled(true);
            }
        }

        this.hooks.configure.call(this);
    }

    public configureTheme(themeData: Partial<ThemeData>) {
        if (!themeData.path) {
            logger(`can not configure theme without defined path: ${JSON.stringify(themeData)}`);
            return;
        }

        const theme = this.themes.find(theme => theme.getData().path === themeData.path);

        if (!theme) {
            logger(`can not find theme with path = '${themeData.path}'`);
            return;
        }

        theme.configure(themeData);

        if (themeData.hasOwnProperty('enabled')) {
            theme.setEnabled(themeData.enabled);
        }
    }

    public enableModule(moduleName: string) {
        const magentoModule = this.modules.find(magentoModule => magentoModule.getName() === moduleName);

        if (!magentoModule) {
            logger(`could not find module by name: ${moduleName}`);
            return;
        }

        magentoModule.setEnabled(true);
    }

    public getThemes() {
        return this.enabledThemes.filter(magentoTheme => magentoTheme.getEnabled());
    }

    public getAllThemes() {
        return this.themes;
    }

    public getModules() {
        return this.modules.filter(magentoModule => magentoModule.getEnabled());
    }

    public getAllModules() {
        return this.modules;
    }

    public getRootDirectory() {
        return this.root;
    }

    public getThemePubDirectories(theme: Theme) {
        return theme.getLocales().map(locale => join(
            this.getStaticDirectory(),
            theme.getData().area,
            theme.getData().path,
            locale
        ));
    }

    public getWebpackOutputDirectories(theme: Theme) {
        if (this.isWebpackDevelopmentMode() || this.experiments.webpack.emitFilesToPub) {
            return theme.getLocales().map(locale => join(
                this.getRootDirectory(),
                ...(
                    this.isWebpackDevelopmentMode()
                    ? ['static']
                    : ['pub', 'static']
                ),
                theme.getData().area,
                theme.getData().path,
                locale
            ));
        }

        return [join(theme.getSourceDirectory(), 'web')];
    }

    public getStaticDirectory() {
        return join(
            this.getRootDirectory(),
            'pub',
            'static'
        );
    }

    public isWebpackDevelopmentMode() {
        return this.webpackDevelopmentMode;
    }

    public setWebpackDevelopmentMode() {
        this.webpackDevelopmentMode = true;
    }

    public setProxyUrl(url: string) {
        this.proxyUrl = url;
    }

    public getProxyUrl(): string {
        if (this.proxyUrl.length === 0) {
            if (!debug.enabled('gpc:logger')) {
                debug.enable('gpc:logger');
            }
            logger(`A project URL must be provided in your projects ${ConfigLoader.CONFIG_FILE}`);
            logger(`\n\nExample:
module.exports = project => {
project.hooks.configure.tap("Set Product Url", () => {
    project.setProxyUrl("https://example.test");
});
}
            `);

            throw new Error(`A project URL must be provided in your projects ${ConfigLoader.CONFIG_FILE}`);
        }

        return this.proxyUrl;
    }

    public setWebpackDevServerConfig(config: Partial<WebpackDevServerConfig>) {
        if (config.proxyUrl) {
            this.proxyUrl = config.proxyUrl;
        }

        this.devServerConfig = config;
    }

    public getWebpackDevServerConfig(): WebpackDevServerConfig {
        return {
            proxyUrl: this.getProxyUrl(),
            devServerUrl: 'green-pistachio.test',
            devServerPort: 8080,
            ...(this.devServerConfig || {})
        };
    }
}
