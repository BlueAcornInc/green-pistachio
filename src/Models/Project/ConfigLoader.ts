import { promises as fs } from 'fs';
import { join } from 'path';
import debug from 'debug';
import glob from 'fast-glob';
import Project from "../Project";
const logger = debug('gpc:configLoader');

export default class ConfigLoader {
    public static CONFIG_FILE = 'green-pistachio.config.js';

    constructor(private project: Project) {}

    get filePath(): string {
        return join(
            this.project.getRootDirectory(),
            ConfigLoader.CONFIG_FILE
        );
    }

    public async configExists(): Promise<boolean> {
        try {
            const configExists = await fs.stat(this.filePath);

            if (configExists) {
                return true;
            }
        } catch (err) {
            logger(`Config file can't be found at ${this.filePath}`);
        }

        return false;
    }

    public async loadConfig() {
        const configExists = await this.configExists();

        if (configExists) {
            try {
                this.interopRequire()(this.project);
            } catch (err) {
                logger(`Problem occurred while running user provided configuration file: ${err}`);
            }
        }

        await this.loadModularConfigFiles();
    }

    private async loadModularConfigFiles() {
        const filePaths = [
            ...this.project.getAllModules(),
            ...this.project.getAllThemes()
        ].map(compilableObject =>
            join(
                compilableObject.getSourceDirectory(),
                ConfigLoader.CONFIG_FILE
            )
        );

        const configFiles = await glob(filePaths);

        for (const configFile of configFiles) {
            try {
                this.interopRequire(configFile)(this.project);
            } catch (err) {
                logger(`Problem occurred while running user provided configuration file: ${err}`);
            }
        }
    }

    private interopRequire(filePath = '') {
        const configFile = filePath || this.filePath;
        const configModule = require(configFile);
        return configModule && configModule.__esModule ? configModule.default : configModule;
    }
}