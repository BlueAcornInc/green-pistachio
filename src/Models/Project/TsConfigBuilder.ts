import { promises as fs } from 'fs';
import { join } from 'path';
import debug from 'debug';
import Project from "../Project";
const logger = debug('gpc:typescript:configBuilder');

export default class TsConfigBuilder {
    public getFilePath(): string {
        return join(__dirname, '..', '..', '..', 'green-pistachio.tsconfig.json');
    }

    public async configFileExists(): Promise<boolean> {
        try {
            await fs.stat(this.getFilePath());

            return true;
        } catch (err) {
            return false;
        }
    }

    public async emitConfigFile(project: Project) {
        logger(`Generating Projects tsconfig.json base`);
        const include: string[] = [];
        const paths: Record<string, string[]> = {}

        for (const module of project.getModules()) {
            const moduleSrcDir = module.getSourceDirectory().replace(project.getRootDirectory(), '');

            const modulePaths: string[] = [
                join(moduleSrcDir, 'view/frontend/web/ts/*'),
                join(moduleSrcDir, 'view/base/web/ts/*'),
                join(moduleSrcDir, 'view/adminhtml/web/ts/*'),
            ];

            paths[`${module.getName()}/*`] = modulePaths.map(modulePath => modulePath.replace(/^\//, ''));

            include.push(join(moduleSrcDir, 'view/frontend/web/ts/**/*'));
            include.push(join(moduleSrcDir, 'view/base/web/ts/**/*'));
            include.push(join(moduleSrcDir, 'view/adminhtml/web/ts/**/*'));
        }

        for (const theme of project.getAllThemes()) {
            const themeSrcDir = theme.getSourceDirectory().replace(project.getRootDirectory(), '');

            include.push(join(themeSrcDir, '**/*'));
            
            // Add theme overrides to module paths
            for (const moduleName of Object.keys(paths)) {
                paths[moduleName].unshift(
                    join(themeSrcDir, moduleName.replace(/\/\*$/, ''), 'web/ts/*').replace(/^\//, '')
                );
            }
        }

        logger(`Resolved include paths to: ${JSON.stringify(include)}`);
        logger(`Resolved module paths to: ${JSON.stringify(paths)}`);

        const writePath = this.getFilePath();
        const config = this.getBaseConfig();

        config.include = include.map(includePath => `./${includePath}`);
        config.compilerOptions.paths = paths;

        try {
            logger(`Writing config file: ${writePath}`);
            await fs.writeFile(writePath, JSON.stringify(config, null, 4));
        } catch (err) {
            logger(`Problem writing config file: ${writePath}`);
            throw err;
        }

        return writePath;
    }

    private getBaseConfig() {
        return {
            "include": [
              "./app/**/*"
            ],
            "compilerOptions": {
              "baseUrl": ".",
              "paths": {},
              "strict": true,
              "esModuleInterop": true,
              "lib": [
                "dom",
                "es2015"
              ],
              "jsx": "react",
              "emitDecoratorMetadata": true,
              "experimentalDecorators": true,
              "target": "es5",
              "allowJs": true,
              "skipLibCheck": true,
              "allowSyntheticDefaultImports": true,
              "forceConsistentCasingInFileNames": true,
              "module": "esnext",
              "moduleResolution": "node",
              "resolveJsonModule": true,
              "isolatedModules": true,
              "noEmit": true
            }
          }
    }
}