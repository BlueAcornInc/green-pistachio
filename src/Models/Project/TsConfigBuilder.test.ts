import mock from 'mock-fs';
import { promises as fs } from 'fs';
import Module from "../Module";
import Project from "../Project";
import Theme from "../Theme";
import TsConfigBuilder from './TsConfigBuilder';

afterEach(() => {
    mock.restore();
});

describe('TS Config Builder', () => {
    it('should generate ts config from project', async () => {
        mock();
        const project = new Project({
            themes: [
                new Theme({
                    area: 'frontend',
                    path: 'Magento/blank',
                    sourceDirectory: 'vendor/magento/theme-frontend-blank'
                }),
                new Theme({
                    area: 'frontend',
                    path: 'Magento/luma',
                    sourceDirectory: 'vendor/magento/theme-frontend-luma'
                }),
                new Theme({
                    area: 'frontend',
                    path: 'BlueAcorn/site',
                    sourceDirectory: 'app/design/frontend/BlueAcorn/site'
                }),
                new Theme({
                    area: 'adminhtml',
                    path: 'Magento/backend',
                    sourceDirectory: 'vendor/magento/theme-adminhtml-backend'
                })
            ],
            modules: [
                new Module({
                    name: 'Magento_Theme',
                    sourceDirectory: 'vendor/magento/module-theme'
                }),
                new Module({
                    name: 'BlueAcorn_Core',
                    sourceDirectory: 'app/code/BlueAcorn/Core'
                })
            ],
            root: '',
        });

        const tsConfigBuilder = new TsConfigBuilder();
        const filePath = await tsConfigBuilder.emitConfigFile(project);

        const fileContent = await fs.readFile(filePath);
        expect(fileContent.toString()).toMatchSnapshot();
    });
});