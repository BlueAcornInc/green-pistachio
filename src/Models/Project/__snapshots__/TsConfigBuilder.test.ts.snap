// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TS Config Builder should generate ts config from project 1`] = `
"{
    \\"include\\": [
        \\"./app/code/BlueAcorn/Core/view/frontend/web/ts/**/*\\",
        \\"./app/code/BlueAcorn/Core/view/base/web/ts/**/*\\",
        \\"./app/code/BlueAcorn/Core/view/adminhtml/web/ts/**/*\\",
        \\"./vendor/magento/theme-frontend-blank/**/*\\",
        \\"./vendor/magento/theme-frontend-luma/**/*\\",
        \\"./app/design/frontend/BlueAcorn/site/**/*\\",
        \\"./vendor/magento/theme-adminhtml-backend/**/*\\"
    ],
    \\"compilerOptions\\": {
        \\"baseUrl\\": \\".\\",
        \\"paths\\": {
            \\"BlueAcorn_Core/*\\": [
                \\"vendor/magento/theme-adminhtml-backend/BlueAcorn_Core/web/ts/*\\",
                \\"app/design/frontend/BlueAcorn/site/BlueAcorn_Core/web/ts/*\\",
                \\"vendor/magento/theme-frontend-luma/BlueAcorn_Core/web/ts/*\\",
                \\"vendor/magento/theme-frontend-blank/BlueAcorn_Core/web/ts/*\\",
                \\"app/code/BlueAcorn/Core/view/frontend/web/ts/*\\",
                \\"app/code/BlueAcorn/Core/view/base/web/ts/*\\",
                \\"app/code/BlueAcorn/Core/view/adminhtml/web/ts/*\\"
            ]
        },
        \\"strict\\": true,
        \\"esModuleInterop\\": true,
        \\"lib\\": [
            \\"dom\\",
            \\"es2015\\"
        ],
        \\"jsx\\": \\"react\\",
        \\"emitDecoratorMetadata\\": true,
        \\"experimentalDecorators\\": true,
        \\"target\\": \\"es5\\",
        \\"allowJs\\": true,
        \\"skipLibCheck\\": true,
        \\"allowSyntheticDefaultImports\\": true,
        \\"forceConsistentCasingInFileNames\\": true,
        \\"module\\": \\"esnext\\",
        \\"moduleResolution\\": \\"node\\",
        \\"resolveJsonModule\\": true,
        \\"isolatedModules\\": true,
        \\"noEmit\\": true
    }
}"
`;
